packages/teleterm/src/ui/DocumentCluster/ClusterResources/useServerSideResources.ts

```diff
@@ -0,0 +1,167 @@
+  const [fetchAttempt, fetch] = useAsync(
+    (startKey: string, filter: AgentFilter) =>
+      retryWithRelogin(ctx, documentUri, () =>
+        fetchFunction({
+          ...filter,
+          limit,
+          clusterUri,
+          startKey,
+        })
+      )
+  );
+
+  // If there is no startKey at the current page's index, there is no more data to get
+  function nextPage() {
+    const proposedKey = keys[pageIndex];
+    if (proposedKey) {
+      return () => {
+        setPageIndex(pageIndex + 1);
+      };
+    }
+    return null;
+  }
+
+  // If we are at the first page (index 0), we cannot fetch more previous data
+  function prevPage() {
+    const newPageIndex = pageIndex - 1;
+    if (newPageIndex > -1) {
+      return () => {
+        setPageIndex(newPageIndex);
+      };
+    }
+    return null;
+  }
+
+  useEffect(() => {
+    const fetchAndUpdateKeys = async () => {
+      const [response, err] = await fetch(keys[pageIndex - 1], agentFilter);
+      // The error will be handled via the fetchAttempt outside.
+      // Return early here as there are no keys to update.
+      if (err) {
+        return;
+      }
+
+      // when we receive data from fetch, we store the startKey (or lack of) according to the current
+      // page index. think of this as "this page's nextKey".
+      // "why don't we just name it nextKey then?"
+      // Mostly because it's called startKey almost everywhere else in the UI, and also because we'd have the same issue
+      // for prevPage if we swapped named, and this comment would be explaining "this page's startKey".
+      const newKeys = [...keys];
+      newKeys[pageIndex] = response.startKey;
+      setKeys(newKeys);
+    };
+    fetchAndUpdateKeys();
+  }, [agentFilter, pageIndex]);
+
+  function updateAgentFilter(filter: AgentFilter) {
+    setPageIndex(0);
+    setAgentFilter(filter);
+  }
+
+  function updateSort(sort: SortType) {
+    updateAgentFilter({ ...agentFilter, sort });
+  }
+
+  function updateSearch(search: string) {
+    updateAgentFilter({ ...agentFilter, query: '', search });
+  }
+
+  function updateQuery(query: string) {
+    updateAgentFilter({ ...agentFilter, search: '', query });
+  }
+
+  function onAgentLabelClick(label: AgentLabel) {
+    const query = addAgentLabelToQuery(agentFilter, label);
+    updateAgentFilter({ ...agentFilter, search: '', query });
+  }
+
+  const pageCount = useMemo(() => {
+    if (fetchAttempt.status !== 'success') {
+      return {
+        from: 0,
+        to: 0,
+        total: 0,
+      };
+    }
+    const from = pageIndex * limit + 1;
+    return {
+      from,
+      to: from + fetchAttempt.data.agentsList.length - 1,
+      total: fetchAttempt.data.totalCount,
+    };
+  }, [fetchAttempt.status]);
+
+  return {
+    fetchAttempt,
+    fetch,
+    updateSearch,
+    updateSort,
+    updateQuery,
+    agentFilter,
+    prevPage: prevPage(),
+    nextPage: nextPage(),
+    onAgentLabelClick,
+    pageCount,
+  };
+}
```

Comment:
I noticed a small issue: when the page is changed, the `page count` becomes **SHOWING 0 - 0 of 0** for a while. It can be fixed this way:
```suggestion 
const pageCount = useMemo(() => {
    const emptyPageCount = {
      from: 0,
      to: 0,
      total: 0,
    };

    if (!(fetchAttempt.data && fetchAttempt.data.totalCount)) {
      return emptyPageCount;
    }

    const startKeyIndex = keys.indexOf(fetchAttempt.data.startKey);
    if (startKeyIndex < 0) {
      return emptyPageCount;
    }

    const from = startKeyIndex * limit + 1;
    return {
      from,
      to: from + fetchAttempt.data.agentsList.length - 1,
      total: fetchAttempt.data.totalCount,
    };
  }, [fetchAttempt, keys]);
```